chapters/unitexpressions.tex

```diff
@@ -89,3 +89,171 @@ \section{The Syntax of Unit Expressions}\label{the-syntax-of-unit-expressions}
+
+How to verify that units are used in compatible ways is current not fully determined by the specification.
+This section gives rules for certain situations, but in general tools are expected to reason according to standard dimensional analysis.
+
+
+\subsection{Standard Dimensional Analysis}\label{standard-dimensional-analysis}
+
+This section gives an incomplete characterization of ``standard dimensional analysis'', also referred to as just \firstuse{dimensional analysis}.
+What is described below applies to unit checking in Modelica -- \emph{dimensional analysis} could have additional meanings in other contexts.
+While Modelica has a concept of \willintroduce{empty units} (described in later sections), standard dimensional analysis only deals with non-empty units such as \lstinline!"m"!, \lstinline!"m/s"!, or \lstinline!"1"!.
+It consists of two parts:
+\begin{itemize}
+\item
+  Unit compatibility requirements.
+\item
+  Rules for deriving the unit of an expression.
+\end{itemize}
+
+Examples of unit compatibility requirements that must be met in Modelica:
+\begin{itemize}
+\item
+  Dimensional homogeneity: The two sides of an equation or assignment statement must have the same unit.
+\item
+  The expression of a declaration equation must have the same unit as the component to which it belongs (special case of dimensional homogeneity).
+\item
+  The two operands of the additive operators \lstinline!+! and \lstinline!-! must have the same unit.
+\item
+  In a function call, the unit of an argument expression must match the unit (unless empty, see below) of the corresponding function input component.
+\end{itemize}
+
+% TODO: Replace these examples by giving unit semantics where operators and built-in functions are defined in the specification,
+% and just include some likes to places where such semantics are given.
+Examples of unit derivation in Modelica:
+\begin{itemize}
+\item
+  The result of additive operators has the same unit as the operands.
+\item
+  The result of multiplication has a unit obtained by multiplying the operands' units.
+\item
+  Built-in operators such as \lstinline!pre! and \lstinline!previous! preserve units, while \lstinline!der! changes the unit by dividing it by \lstinline!"s"!.
+\end{itemize}
+
+
+\subsection{Empty and Undefined Units}\label{empty-and-undefined-units}
+
+In situations where the specification does not prescribe how to determine the unit of an expression, the unit of that expression is said to be \firstuse[undefined unit]{undefined}.
+It is then not possible for a tool to reject the equation (or other construct) with support in the specification, and options for the tool include silently not performing unit checking, or applying checks based dimensional analysis.
+
+The unit of an expression can also be defined as being \firstuse[empty unit]{empty}, see below.
+In certain places (see below), expressions with empty unit can be implicitly cast to suitable units.
```

Comment:
I'm not a friend of "soft" qualifiers like "certain", "some of", etc. in specification texts. I think it would be better to enumerate all the places/sections/cases you mean to include in this statement.

Reply:
The paragraph is part of the section introduction, so it isn't meant to convey the details.  I agree that later, when the actual rule is given, one has to avoid formulations that risk causing unnecessary ambiguity.

Does it work better now, with the reformulated presentation of the corresponding list of cases?  (I don't think this list needs a fall-through case, as it should always be possible to add more cases in a backwards compatible way.)

Reply:
> The paragraph is part of the section introduction, so it isn't meant to convey the details. 

Should it then be marked non-normative? Otherwise, it's normative, right? (I'm not too firm on the procedural details of this)

> reformulated presentation of the corresponding list of cases?

I can't tell which list you mean here?

> I don't think this list needs a fall-through case, as it should always be possible to add more cases in a backwards compatible way.)

This is not a contradiction! You can first specify (random example: operators covered by some procedure):
* Addition: do A
* Subtraction: do B
* All others: undefined

and then later refine by adding more cases
* Addition: do A
* Subtraction: do B
* Multiplication: do C
* Division: do D
* All others: undefined

The set of the fallthrough/uncovered cases shrinks by adding more cases, but it's always crystal clear which ones are "undefined" (first: everything except +, then everything except +,-,*,/)

Reply:
> > The paragraph is part of the section introduction, so it isn't meant to convey the details.
> 
> Should it then be marked non-normative? Otherwise, it's normative, right? (I'm not too firm on the procedural details of this)

I'm not sure.  It's a matter of style where we have at least decided to not explicitly mark chapter introductions as non-normative even though their nature is non-normative.  In the case at hand, I'd say that the _(see below)_ should be sufficient to ensure the reader that there's a specific meaning of _certain places_ to be found by looking at the sections below.

Reply:
> > reformulated presentation of the corresponding list of cases?
> 
> I can't tell which list you mean here?

I refer to formulations like this one:
> When encountering the empty unit in the following situations, the inferred unit is (uniquely) determined by ensuring that unit compatibility requirements are fulfilled:

Reply:
> > I don't think this list needs a fall-through case, as it should always be possible to add more cases in a backwards compatible way.)
> 
> This is not a contradiction! You can first specify (random example: operators covered by some procedure):

The nice thing about this list (_When encountering the empty unit in…_) is that it doesn't leave any situation undefined.  In the other cases, the unit isn't determined by the context, but always becomes `"1"` as (now) described directly below the list.